File: 167/two-sum-ii-input-array-is-sorted.js
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function (numbers, target) {
    let hash = {};
    for (let i = 0, len = numbers.length; i < len; i++) {
        let item = target - numbers[i];
        if (hash[item] === undefined) {
            hash[numbers[i]] = i;
            continue;
        }
        return [hash[item] + 1, i + 1];
    }
};

var twoSum = function (numbers, target) {
    let i = 1, j = numbers.length;
    while (i < j) {
        if (numbers[i - 1] + numbers[j - 1] === target) {
            return [i, j];
        } else if (numbers[i - 1] + numbers[j - 1] > target) {
            j--;
        } else {
            i++;
        }
    }
};

File: 167/TwoSumIIInputArrayIsSorted.scala
object TwoSumIIInputArrayIsSorted {
  def twoSum(numbers: Array[Int], target: Int): Array[Int] = {
    @annotation.tailrec
    def twoSum(low: Int, high: Int): Array[Int] = {
      val s = numbers(low) + numbers(high)
      if (s == target)
        Array(low+1, high+1)
      else if (s < target)
        twoSum(low+1, high)
      else
        twoSum(low, high-1)
    }

    twoSum(0, numbers.length-1)
  }

  def main(args: Array[String]): Unit = {
    val numbers = Array(2, 7, 11, 15)
    val target = 9
    for (item <- twoSum(numbers, target)) { println(item) }
  }
}
